// rsdi/resolvers/ObjectResolver.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var AbstractResolver_1 = __importDefault(require("./AbstractResolver"));
var errors_1 = require("../errors");
var DIContainer_1 = require("../DIContainer");
/**
 * ObjectDefinition creates objects from the provided class.
 *
 */
var ObjectResolver = /** @class */ (function (_super) {
    __extends(ObjectResolver, _super);
    function ObjectResolver(constructorFunction) {
        var _this = _super.call(this) || this;
        _this.deps = [];
        _this.methods = [];
        _this.resolve = function (diContainer, parentDeps) {
            var _a;
            if (parentDeps === void 0) { parentDeps = []; }
            var constructorParameters = (0, DIContainer_1.resolveFunctionParameters)(diContainer, _this.deps, parentDeps);
            var object = new ((_a = _this.constructorFunction).bind.apply(_a, __spreadArray([void 0], constructorParameters, false)))();
            _this.methods.forEach(function (method) {
                var methodName = method.methodName, args = method.args;
                if (object[methodName] === undefined) {
                    throw new errors_1.MethodIsMissingError(object.constructor.name, methodName);
                }
                var resolvedArgs = (0, DIContainer_1.resolveFunctionParameters)(diContainer, args, parentDeps);
                object[methodName].apply(object, resolvedArgs);
            });
            return object;
        };
        if (typeof constructorFunction !== "function") {
            throw new errors_1.InvalidConstructorError();
        }
        _this.constructorFunction = constructorFunction;
        return _this;
    }
    /**
     * Defines constructor parameters for a given class.
     * @param deps
     */
    ObjectResolver.prototype.construct = function () {
        var deps = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            deps[_i] = arguments[_i];
        }
        this.deps = deps;
        return this;
    };
    /**
     * After DIContainer constructs object of a given class, DIContainer calls additional
     * `method`-s of a given class.
     *
     * @param methodName - string, name of a given class
     * @param args - match method signature of a given class method
     */
    ObjectResolver.prototype.method = function (methodName) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.methods.push({
            methodName: methodName,
            args: args,
        });
        return this;
    };
    return ObjectResolver;
}(AbstractResolver_1.default));
exports.default = ObjectResolver;
